packages/theme-data/src/utils/extendTheme.js

```diff
@@ -6,8 +6,21 @@
  * Add additional roles to a theme
  */
 export default function extendTheme(theme, extension) {
+  // NOTE: We do *not* use Object spread in `reduce`, since it copies the entire
+  // object in every iteration and becomes a performance bottleneck. Instead, we
+  // copy `theme` once at the beginning and then apply changes to that same
+  // object instance `acc` using property assignment and `Object.assign`.
   return Object.keys(extension).reduce(
-    (acc, key) => ({ ...acc, [key]: { ...theme[key], ...extension[key] } }),
-    theme
+    (acc, key) => {
+      if (!acc[key]) {
+        acc[key] = extension[key];
+      } else {
+        // Use Object spread here so that we create a copy of `acc[key]` and do {
+        // not mutate an object that is also referenced by `theme`.
+        acc[key] = { ...acc[key], ...extension[key] };
```

Comment:
@nemoDreamer originally I was doing:

```jsx
acc[key] = Object.assign(acc[key], extension[key]);
```

`acc[key]` in this case was a reference to an object that was shared with the original `theme` argument, so I was mutating a value used outside of `extendTheme`. I (think I) fixed it by using Object spread to create a copy of that object before applying the extension.

Reply:
You could also use `Object.assign({}, acc[key], extension[key])` here to apply everything to a new object rather than modify `acc[key]`.

Reply:
I like that better! Pushed a change